File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char *s) {
    int lastIndex[128];
    memset(lastIndex, -1, sizeof(lastIndex));

    int left = 0, maxLen = 0;
    int n = strlen(s);

    for (int right = 0; right < n; right++) {
        unsigned char ch = s[right];
        if (lastIndex[ch] >= left) {
            left = lastIndex[ch] + 1;
        }
        lastIndex[ch] = right;
        int windowLen = right - left + 1;
        if (windowLen > maxLen) {
            maxLen = windowLen;
        }
    }

    return maxLen;
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_index = {}  # last seen index of each character
        left = 0
        max_len = 0

        for right, ch in enumerate(s):
            if ch in last_index and last_index[ch] >= left:
                left = last_index[ch] + 1
            last_index[ch] = right
            max_len = max(max_len, right - left + 1)

        return max_len